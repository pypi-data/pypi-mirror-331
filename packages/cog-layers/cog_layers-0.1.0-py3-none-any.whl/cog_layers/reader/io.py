import typing
import functools

import obstore as obs


RangeRequestFuncType = typing.Callable[
    [str, str, int, int, typing.Any], typing.Awaitable[bytes]
]


@functools.lru_cache(maxsize=1)
def _get_default_obstore_client(bucket_name: str, **kwargs) -> obs.store.S3Store:
    """Default obstore client with a singleton cache."""
    return obs.store.S3Store(bucket_name, **kwargs)


async def send_range_obstore(
    bucket: str, key: str, start: int, end: int, client: typing.Any | None = None
) -> bytes:
    """Send a range request with obstore.

    Creates a single client and reuses unless overridden by caller.
    """
    client = client or _get_default_obstore_client(bucket)
    r = await obs.get_range_async(client, key, start=start, end=end)
    return r.to_bytes()
