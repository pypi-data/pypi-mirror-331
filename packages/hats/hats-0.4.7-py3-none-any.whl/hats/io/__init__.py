"""Utilities for reading and writing catalog files"""

from .parquet_metadata import (
    read_row_group_fragments,
    row_group_stat_single_value,
    write_parquet_metadata,
    write_parquet_metadata_for_batches,
)
from .paths import (
    get_common_metadata_pointer,
    get_parquet_metadata_pointer,
    get_partition_info_pointer,
    get_point_map_file_pointer,
    pixel_catalog_file,
    pixel_directory,
)
