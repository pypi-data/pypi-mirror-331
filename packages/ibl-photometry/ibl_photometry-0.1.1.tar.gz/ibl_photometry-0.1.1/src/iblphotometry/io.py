import numpy as np
import pandas as pd
from pathlib import Path
import warnings
import pandera
from typing import Optional

from iblphotometry.neurophotometrics import (
    LIGHT_SOURCE_MAP,
    LED_STATES,
)


def from_raw_neurophotometrics_file_to_raw_df(
    path: str | Path, validate=True, version='new'
) -> pd.DataFrame:
    """reads in a file as generated by the neurophotometrics FP3002 (both new and old versions) with validation

    Args:
        path (str | Path): path to the file, can be in either .csv or .pqt format
        validate (bool, optional): If True, validates the file. Defaults to True.
        version (str, optional): 'new' or 'old' version of the neurophotometrics file format. Defaults to 'new'.

    Returns:
        pd.DataFrame: the data as a raw dataframe format
    """
    path = Path(path) if isinstance(path, str) else path
    match path.suffix:
        case '.csv':
            raw_df = pd.read_csv(path)
        case '.pqt':
            raw_df = pd.read_parquet(path)

    if validate:
        if version == 'old':
            if 'Flags' in raw_df.columns:
                raw_df = raw_df.rename(columns={'Flags': 'LedState'})
        raw_df = validate_neurophotometrics_df(raw_df, version=version)

    return raw_df


def from_raw_neurophotometrics_df_to_ibl_df(
    raw_df: pd.DataFrame, rois=None, drop_first=True
) -> pd.DataFrame:
    """reads in a dataframe with the raw photometry data as generated by the neurophotometrics FP3002 into the ibl photometry dataformat.


    Args:
        raw_df (pd.DataFrame): as returned by `from_raw_neurophotometrics_file_to_raw_df`
        rois (_type_, optional): names of the rois as selected by the user in the acquisition UI. If None, the names are inferred from the data. Defaults to None.
        drop_first (bool, optional): Drop the The first frame, which has all LEDs on by default. Defaults to True.

    Returns:
        pd.DataFrame: the data in the ibl photometry data format
    """
    if rois is None:
        rois = infer_data_columns(raw_df)

    ibl_df = raw_df.filter(items=rois, axis=1).sort_index(axis=1)
    timestamp_name = (
        'SystemTimestamp' if 'SystemTimestamp' in raw_df.columns else 'Timestamp'
    )
    ibl_df['times'] = raw_df[timestamp_name]
    ibl_df['wavelength'] = np.nan
    ibl_df['name'] = ''
    ibl_df['color'] = ''

    # TODO the names column in channel_meta_map should actually be user defined (experiment description file?)
    channel_meta_map = pd.DataFrame(LIGHT_SOURCE_MAP)
    led_states = pd.DataFrame(LED_STATES).set_index('Condition')

    # downward compatibility fix for older than old format
    if 'Flags' in raw_df.columns:
        raw_df = raw_df.rename(columns={'Flags': 'LedState'})
    states = raw_df['LedState']

    for state in states.unique():
        ir, ic = np.where(led_states == state)
        # if not present, multiple LEDs are active
        if ic.size == 0:
            # find row
            ir = np.argmax(led_states['No LED ON'] > state) - 1
            # find active combo
            possible_led_combos = [(1, 2), (1, 3), (2, 3), (1, 2, 3)]
            for combo in possible_led_combos:  # drop enumerate
                if state == sum([led_states.iloc[ir, c] for c in combo]):
                    name = '+'.join([channel_meta_map['name'][c] for c in combo])
                    color = '+'.join([channel_meta_map['color'][c] for c in combo])
                    wavelength = np.nan
                    ibl_df.loc[states == state, ['name', 'color', 'wavelength']] = (
                        name,
                        color,
                        wavelength,
                    )
        else:
            for cn in ['name', 'color', 'wavelength']:
                ibl_df.loc[states == state, cn] = channel_meta_map.iloc[ic[0]][cn]

    # drop first frame
    if drop_first:
        ibl_df = ibl_df.iloc[1:].reset_index()

    return ibl_df


def from_raw_neurophotometrics_file_to_ibl_df(
    path: str | Path, drop_first=True, validate=True, version='new'
) -> pd.DataFrame:
    """convenience function that chains `from_raw_neurophotometrics_file_to_raw_df` and `from_raw_neurophotometrics_df_to_ibl_df`. See docstrings

    Args:
        path (str | Path): _description_
        drop_first (bool, optional): Drop the The first frame, which has all LEDs on by default. Defaults to True.
        validate (bool, optional): If True, validates the file. Defaults to True.
        version (str, optional): 'new' or 'old' version of the neurophotometrics file format. Defaults to 'new'.

    Returns:
        pd.DataFrame: _description_
    """
    raw_df = from_raw_neurophotometrics_file_to_raw_df(
        path, validate=validate, version=version
    )
    ibl_df = from_raw_neurophotometrics_df_to_ibl_df(raw_df, drop_first=drop_first)

    return ibl_df


def from_ibl_pqt_to_ibl_df(path: str | Path, validate=False):
    ibl_df = pd.read_parquet(path)
    if validate is True:
        ibl_df = validate_ibl_dataframe(ibl_df)
    return ibl_df


def from_ibl_dataframe(
    ibl_df: pd.DataFrame,
    data_columns: list[str] | None = None,
    time_column: str | None = None,
    channel_column: str = 'name',
    channel_names: list[str] | None = None,
    rename: dict | None = None,
    validate: bool = True,
) -> dict:
    """main function to convert to analysis ready format


    Args:
        ibl_df (pd.DataFrame): the dataframe, as stored in the photometry.signal.pqt
        data_columns (list[str], optional): The names of the columns in the dataframe that contain the signals of different fibers. By default, they are named RegionXX. If None is provided, All columns that start with `Region` or `G` are treated as data columns. Defaults to None.
        time_column (str, optional): The name of the column that contains the timestamps. If None is provided, it is assumed that `time` is in the name. Defaults to None.
        channel_column (str, optional): The name of the column that contains. Defaults to 'name'.
        channel_names (list[str], optional): The names of the acquisition channel / frequency bands that are acquired. Defaults to None.
        rename (dict, optional): a renaming map that maps the names of the columns to brain areas. Example: {'RegionXX':'DMS'}. Defaults to None.

    Returns:
        dict: A dict with the keys being the names of the acquisition channels, the values being nap.TsdFrames with the columns containing the data of the different fibers
    """
    # from a raw dataframe as it is stored in ONE (signal.pqt)
    # data_columns is a list of str that specifies the names of the column that hold the actual data, like 'RegionXX'
    # channel_column is the column that specifies the temporally multiplexed acquisition channels

    if validate:
        ibl_df = validate_ibl_dataframe(ibl_df)

    data_columns = infer_data_columns(ibl_df) if data_columns is None else data_columns

    # infer name of time column if not provided
    if time_column is None:
        time_columns = [col for col in ibl_df.columns if 'time' in col.lower()]
        assert len(time_columns) == 1
        time_column = time_columns[0]

    # infer channel names if they are not explicitly provided
    if channel_names is None:
        channel_names = ibl_df[channel_column].unique()

    # drop empty acquisition channels
    to_drop = ['None', '']
    channel_names = [ch for ch in channel_names if ch not in to_drop]

    dfs = {}
    for channel in channel_names:
        # get the data for the band
        df = ibl_df.groupby(channel_column).get_group(channel)
        # if rename dict is passed, rename Region0X to the corresponding brain region
        if rename is not None:
            df = df.rename(columns=rename)
            data_columns = rename.values()
        dfs[channel] = df.set_index(time_column)[data_columns]

    return dfs


def from_ibl_pqt(
    signal_pqt_path: str | Path,
    locations_pqt_path: Optional[str | Path] = None,
    validate=True,
):
    """reads in photometry data stored in the ibl format as a .pqt file. If provided, uses the metadata stored in the locations.pqt file as well.

    Args:
        signal_pqt_path (str | Path): _description_
        locations_pqt_path (Optional[str  |  Path], optional): _description_. Defaults to None.

    Returns:
        _type_: _description_
    """

    ibl_df = from_ibl_pqt_to_ibl_df(signal_pqt_path, validate=validate)
    if locations_pqt_path is not None:
        locations_df = pd.read_parquet(locations_pqt_path)
        return from_ibl_dataframes(ibl_df, locations_df)
    else:
        warnings.warn(
            'loading a photometry.signal.pqt file without its corresponding photometryROI.locations.pqt'
        )
        data_columns = None
        rename = None

    read_config = dict(
        data_columns=data_columns,
        time_column='times',
        channel_column='name',
        rename=rename,
    )

    return from_ibl_dataframe(ibl_df, **read_config)


def from_ibl_dataframes(ibl_df: pd.DataFrame, locations_df: pd.DataFrame):
    # if locations are present
    data_columns = (list(locations_df.index),)
    rename = locations_df['brain_region'].to_dict()

    read_config = dict(
        data_columns=data_columns,
        time_column='times',
        channel_column='name',
        rename=rename,
    )

    return from_ibl_dataframe(ibl_df, **read_config)


def from_raw_neurophotometrics_file(
    path: str | Path, drop_first=True, validate=True, version='new'
) -> dict:
    """reads in a file generated by the neurophotometrics FP3002 into the analysis ready format

    Args:
        path (str | Path): _description_
        drop_first (bool, optional): Drop the The first frame, which has all LEDs on by default. Defaults to True.
        validate (bool, optional): If True, validates the file. Defaults to True.
        version (str, optional): 'new' or 'old' version of the neurophotometrics file format. Defaults to 'new'.

    Returns:
        dict: _description_
    """
    ibl_df = from_raw_neurophotometrics_file_to_ibl_df(
        path, drop_first=drop_first, validate=validate, version=version
    )
    # data_columns = infer_data_columns(ibl_df) if data_columns is None else data_columns
    read_config = dict(
        # data_columns=data_columns,
        time_column='times',
        channel_column='name',
    )
    return from_ibl_dataframe(ibl_df, **read_config)


def read_digital_inputs_file(path: str | Path, validate=True) -> pd.DataFrame:
    if path.suffix == '.csv':
        return read_digital_inputs_csv(path, validate=validate)
    if path.suffix == '.pqt':
        df_digital_inputs = pd.read_parquet(path)
    if validate:
        df_digital_inputs = validate_neurophotometrics_digital_inputs(df_digital_inputs)
    return df_digital_inputs


def read_digital_inputs_csv(path: str | Path, validate=True) -> pd.DataFrame:
    df_digital_inputs = pd.read_csv(path, header=None)
    df_digital_inputs.columns = [
        'ChannelName',
        'Channel',
        'AlwaysTrue',
        'SystemTimestamp',
        'ComputerTimestamp',
    ]
    if validate:
        df_digital_inputs = validate_neurophotometrics_digital_inputs(df_digital_inputs)
    return df_digital_inputs


"""
##     ##    ###    ##       #### ########     ###    ######## ####  #######  ##    ##
##     ##   ## ##   ##        ##  ##     ##   ## ##      ##     ##  ##     ## ###   ##
##     ##  ##   ##  ##        ##  ##     ##  ##   ##     ##     ##  ##     ## ####  ##
##     ## ##     ## ##        ##  ##     ## ##     ##    ##     ##  ##     ## ## ## ##
 ##   ##  ######### ##        ##  ##     ## #########    ##     ##  ##     ## ##  ####
  ## ##   ##     ## ##        ##  ##     ## ##     ##    ##     ##  ##     ## ##   ###
   ###    ##     ## ######## #### ########  ##     ##    ##    ####  #######  ##    ##
"""


def validate_ibl_dataframe(ibl_df: pd.DataFrame, data_columns=None) -> pd.DataFrame:
    data_columns = infer_data_columns(ibl_df) if data_columns is None else data_columns
    schema_ibl_data = pandera.DataFrameSchema(
        columns=dict(
            times=pandera.Column(pandera.Float64),
            # valid=pandera.Column(pandera.Bool), # optionally present
            wavelength=pandera.Column(pandera.Float64, nullable=True),
            name=pandera.Column(pandera.String),
            color=pandera.Column(pandera.String),
            **{k: pandera.Column(pandera.Float64) for k in data_columns},
        )
    )
    return schema_ibl_data.validate(ibl_df)


def validate_neurophotometrics_df(
    df: pd.DataFrame,
    data_columns=None,
    version='new',  # or 'old' - TODO to be replaced
) -> pd.DataFrame:
    data_columns = infer_data_columns(df) if data_columns is None else data_columns

    match version:
        case 'new':  # kcenia, carolina
            schema_raw_data = pandera.DataFrameSchema(
                columns=dict(
                    FrameCounter=pandera.Column(pandera.Int64),
                    SystemTimestamp=pandera.Column(pandera.Float64),
                    LedState=pandera.Column(pandera.Int16, coerce=True),
                    ComputerTimestamp=pandera.Column(pandera.Float64),
                    **{k: pandera.Column(pandera.Float64) for k in data_columns},
                )
            )
        case 'old':  # alejandro
            schema_raw_data = pandera.DataFrameSchema(
                columns=dict(
                    FrameCounter=pandera.Column(pandera.Int64),
                    Timestamp=pandera.Column(pandera.Float64),
                    LedState=pandera.Column(pandera.Int16, coerce=True),
                    **{k: pandera.Column(pandera.Float64) for k in data_columns},
                )
            )

    return schema_raw_data.validate(df)


def validate_neurophotometrics_digital_inputs(df: pd.DataFrame) -> pd.DataFrame:
    schema_digital_inputs = pandera.DataFrameSchema(
        columns=dict(
            ChannelName=pandera.Column(str, coerce=True),
            Channel=pandera.Column(pandera.Int8, coerce=True),
            AlwaysTrue=pandera.Column(bool, coerce=True),
            SystemTimestamp=pandera.Column(pandera.Float64),
            ComputerTimestamp=pandera.Column(pandera.Float64),
        )
    )
    return schema_digital_inputs.validate(df)


def infer_data_columns(df: pd.DataFrame) -> list[str]:
    # this hacky parser currently deals with the inconsistency between carolinas and alejandros extraction
    # https://github.com/int-brain-lab/ibl-photometry/issues/35
    data_columns = [
        col for col in df.columns if col.startswith('Region') or col.startswith('G')
    ]
    return data_columns
