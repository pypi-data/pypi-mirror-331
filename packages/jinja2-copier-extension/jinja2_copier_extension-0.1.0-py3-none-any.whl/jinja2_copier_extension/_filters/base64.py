"""Jinja2 filters for Base64 encoding/decoding."""

from __future__ import annotations

from base64 import b64decode
from base64 import b64encode

__all__ = ["do_b64decode", "do_b64encode"]


def do_b64decode(value: str, encoding: str = "utf-8") -> str:
    """Decode a Base64 encoded string.

    Args:
        value: A Base64 encoded string.
        encoding: The encoding with which to decode the Base64 decoded bytes.

    Returns:
        The Base64 decoded string.
    """
    return b64decode(value).decode(encoding)


def do_b64encode(value: str, encoding: str = "utf-8") -> str:
    """Encode a string using Base64.

    Args:
        value: A string.
        encoding: The encoding with which to encode the string.

    Returns:
        The Base64 encoded string.
    """
    return b64encode(value.encode(encoding)).decode()
