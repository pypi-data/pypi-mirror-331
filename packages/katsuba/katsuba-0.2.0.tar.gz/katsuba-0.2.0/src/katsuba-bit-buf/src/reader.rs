use std::{io, marker::PhantomData, ptr, slice};

// The maximum number of bits that can be stored in lookahead.
//
// We target to have an amount between 56 and 63 bits in the
// buffer. Since we only refill by whole bytes, it means the
// low 3 bits never change.
const BUFFER_SIZE: u32 = u64::BITS - 1;

// The maximum number of bits to be consumed after a refill.
//
// Since we refill by whole bytes only, this is the smallest
// value where a whole byte doesn't fit in anymore.
const CONSUMABLE_BITS: u32 = BUFFER_SIZE & !7;

#[inline(always)]
unsafe fn read_64_le(ptr: *const u8) -> u64 {
    // SAFETY: Caller must provide a valid pointer.
    let value = unsafe { ptr::read_unaligned(ptr.cast::<u64>()) };
    value.to_le()
}

// There are no stable branch prediction hints for us.
// We emulate them by nudging the compiler into optimizing
// around the paths which do not call this function.
#[cold] // :)
#[inline(always)]
fn cold() {}

/// A buffer which enables bit-based deserialization of data.
///
/// Individual bit reading starts at the LSB of the byte, working
/// towards the MSB.
///
/// When reading bits, users must manually call [`Self::refill_bits`]
/// first and do appropriate checks based on how many bits are left.
///
/// When wanting to read from full byte boundaries with some stale
/// buffered bits, [`Self::invalidate_and_realign_ptr`] can help.
#[derive(Debug)]
pub struct BitReader<'a> {
    // Pointer to the next byte where the bit lookahead
    // buffer will be fetched from.
    ptr: *const u8,

    // A marker pointer past which it is not safe anymore
    // to take the fast refill path which blindly copies.
    safeguard: *const u8,

    // One-past-the-end pointer in the spanned byte view.
    end: *const u8,

    // The pre-fetched lookahead bufer to extract bits at.
    lookahead: u64,

    // The number of bits available for consumption from
    // the `lookahead` buffer.
    remaining: u32,

    // Marker to teach borrowchk this is holding onto a slice.
    _m: PhantomData<&'a [u8]>,
}

impl<'a> BitReader<'a> {
    /// Creates a new [`BitReader`] over a given byte slice.
    pub const fn new(data: &'a [u8]) -> Self {
        let (ptr, len) = (data.as_ptr(), data.len());

        // SAFETY: All pointer arithmetic in bounds or one past the end.
        unsafe {
            Self {
                ptr,
                safeguard: ptr.add(len.saturating_sub(7)),
                end: ptr.add(len),
                lookahead: 0,
                remaining: 0,
                _m: PhantomData,
            }
        }
    }

    #[inline(always)]
    fn can_read_in_fast_path(&self) -> bool {
        self.ptr < self.safeguard
    }

    // Gets the remaining untouched bytes in the reader.
    #[inline]
    pub fn untouched_bytes(&self) -> usize {
        // SAFETY: Byte pointers are derived from the same object,
        // with `ptr <= end` being an internally maintained invariant.
        unsafe { self.end.offset_from(self.ptr) as usize }
    }

    /// Gets the total number of remaining bits in the reader.
    #[inline]
    pub fn remaining_bits(&self) -> usize {
        (self.untouched_bytes() << 3) + self.remaining as usize
    }

    /// Gets the bits currently buffered in the reader.
    #[inline]
    pub fn buffered_bits(&self) -> u32 {
        self.remaining
    }

    /// Invalidates the current bit lookahead and resets the pointer
    /// back to the first untouched byte.
    ///
    /// Untouched in this case means no partial bit reads overlapping
    /// with the memory region of a byte have happened yet.
    #[inline(always)]
    pub fn realign_to_byte(&mut self) {
        // SAFETY: Decrementing the pointer is fine since we move within
        // a fraction of the increment done by a refill operation.
        self.ptr = unsafe { self.ptr.sub(self.remaining as usize >> 3) };

        self.lookahead = 0;
        self.remaining = 0;
    }

    // SAFETY: Caller must make sure `self.ptr` is valid for 8 byte read.
    #[inline]
    unsafe fn refill_branchless(&mut self) {
        // Read from current bit pointer and prefill the entire lookahead.
        self.lookahead |= unsafe { read_64_le(self.ptr) } << self.remaining;

        // Advance the read cursor for the next refill.
        //
        // This code seemingly increases the number of instructions
        // from 3 (sub, shr, add) to 4 (add, shr, and, sub), but with
        // dedicated bitfield extraction support it stays at 3.
        //
        // Note that the dependency chain decreases from 3 to 2 however,
        // which may result in higher throughput.
        unsafe {
            self.ptr = self.ptr.add(CONSUMABLE_BITS as usize >> 3);
            self.ptr = self.ptr.sub((self.remaining as usize >> 3) & 7);
        }

        // Update bit count to reflect full buffer.
        self.remaining |= CONSUMABLE_BITS;
    }

    #[inline]
    fn refill_slow(&mut self) {
        while self.remaining < CONSUMABLE_BITS {
            if self.ptr == self.end {
                cold();
                break;
            }

            // SAFETY: `ptr` hasn't reached the buffer end yet.
            unsafe {
                self.lookahead |= (*self.ptr as u64) << self.remaining;
                self.ptr = self.ptr.add(1);
            }

            self.remaining += u8::BITS;
        }
    }

    /// Refills the bit lookahead buffer and returns the number of
    /// available bits for consumption.
    ///
    /// When this buffer is exhausted, another refill must be done.
    pub fn refill_bits(&mut self) -> u32 {
        debug_assert!(self.ptr <= self.end);
        debug_assert!(self.remaining <= BUFFER_SIZE);

        if self.can_read_in_fast_path() {
            // SAFETY: We have enough bytes for an unchecked refill.
            unsafe {
                self.refill_branchless();
            }
        } else {
            cold();
            self.refill_slow();
        }

        self.remaining
    }

    /// Returns the next `count` bits from the internal buffer without removing
    /// them, if available.
    #[inline]
    pub fn peek(&mut self, count: u32) -> io::Result<u64> {
        if count <= CONSUMABLE_BITS && count <= self.remaining {
            Ok(self.lookahead & ((1 << count) - 1))
        } else {
            Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "attempted to read out of bounds",
            ))
        }
    }

    /// Removes `count` bits from the internal buffer, if available.
    #[inline]
    pub fn consume(&mut self, count: u32) -> io::Result<()> {
        if count <= self.remaining {
            self.lookahead >>= count;
            self.remaining -= count;

            Ok(())
        } else {
            Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "attempted to read out of bounds",
            ))
        }
    }

    /// Reads `nbytes` raw bytes from the internal buffer.
    ///
    /// These are borrowed from the underlying byte view without
    /// copying them.
    pub fn read_bytes(&mut self, count: usize) -> io::Result<&'a [u8]> {
        if count <= self.untouched_bytes() {
            // SAFETY: A bounds check was done and an appropriate lifetime is
            // inferred through the function signature.
            unsafe {
                let value = slice::from_raw_parts(self.ptr, count);
                self.ptr = self.ptr.add(count);
                Ok(value)
            }
        } else {
            Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "attempted to read out of bounds",
            ))
        }
    }
}

// SAFETY: The `BitReader` API does not expose the internal pointers
// in ways which can compromise Rust's safety guarantees.
unsafe impl Send for BitReader<'_> {}
unsafe impl Sync for BitReader<'_> {}
