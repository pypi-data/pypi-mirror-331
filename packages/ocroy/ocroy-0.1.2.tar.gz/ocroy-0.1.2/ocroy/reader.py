from __future__ import annotations

from pathlib import Path


def read_image(image_path: str | Path) -> bytes:
    with open(image_path, "rb") as fb:
        return fb.read()
