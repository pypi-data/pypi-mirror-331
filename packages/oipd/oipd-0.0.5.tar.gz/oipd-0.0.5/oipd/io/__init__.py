from .csv_reader import CSVReader
from .dataframe_reader import DataFrameReader

__all__ = [
    "CSVReader",
    "DataFrameReader",
]
