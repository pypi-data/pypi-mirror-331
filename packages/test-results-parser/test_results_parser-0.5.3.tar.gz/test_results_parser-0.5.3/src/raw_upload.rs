use anyhow::Context;

use base64::prelude::*;
use pyo3::prelude::*;
use std::collections::HashSet;
use std::io::prelude::*;

use flate2::bufread::ZlibDecoder;

use quick_xml::reader::Reader;
use serde::Deserialize;

use crate::junit::{get_position_info, use_reader};
use crate::testrun::ParsingInfo;

#[derive(Deserialize, Debug, Clone)]
struct TestResultFile {
    filename: String,
    data: String,
}
#[derive(Deserialize, Debug, Clone)]
struct RawTestResultUpload {
    #[serde(default)]
    network: Option<HashSet<String>>,
    test_results_files: Vec<TestResultFile>,
}

#[derive(Debug, Clone)]
struct ReadableFile {
    filename: String,
    data: Vec<u8>,
}

const LEGACY_FORMAT_PREFIX: &[u8] = b"# path=";
const LEGACY_FORMAT_SUFFIX: &[u8] = b"<<<<<< EOF";

fn serialize_to_legacy_format(readable_files: Vec<ReadableFile>) -> Vec<u8> {
    let mut res = Vec::new();
    for file in readable_files {
        res.extend_from_slice(LEGACY_FORMAT_PREFIX);
        res.extend_from_slice(file.filename.as_bytes());
        res.extend_from_slice(b"\n");
        res.extend_from_slice(&file.data);
        res.extend_from_slice(b"\n");
        res.extend_from_slice(LEGACY_FORMAT_SUFFIX);
        res.extend_from_slice(b"\n");
    }
    res
}

#[pyfunction]
#[pyo3(signature = (raw_upload_bytes))]
pub fn parse_raw_upload(raw_upload_bytes: &[u8]) -> anyhow::Result<(Vec<ParsingInfo>, Vec<u8>)> {
    let upload: RawTestResultUpload =
        serde_json::from_slice(raw_upload_bytes).context("Error deserializing json")?;
    let network: Option<HashSet<String>> = upload.network;

    let mut results: Vec<ParsingInfo> = Vec::with_capacity(upload.test_results_files.len());
    let mut readable_files: Vec<ReadableFile> = Vec::with_capacity(upload.test_results_files.len());

    for file in upload.test_results_files {
        let decoded_file_bytes = BASE64_STANDARD
            .decode(file.data)
            .context("Error decoding base64")?;

        let mut decoder = ZlibDecoder::new(decoded_file_bytes.as_slice());

        let mut decompressed_file_bytes = Vec::new();
        decoder
            .read_to_end(&mut decompressed_file_bytes)
            .context("Error decompressing file")?;

        let mut reader = Reader::from_reader(decompressed_file_bytes.as_slice());
        reader.config_mut().trim_text(true);
        let reader_result = use_reader(&mut reader, network.as_ref()).with_context(|| {
            let pos = reader.buffer_position();
            let (line, col) = get_position_info(&decompressed_file_bytes, pos.try_into().unwrap());
            format!(
                "Error parsing JUnit XML in {} at {}:{}",
                file.filename, line, col
            )
        })?;

        results.push(reader_result);

        let readable_file = ReadableFile {
            data: decompressed_file_bytes,
            filename: file.filename,
        };
        readable_files.push(readable_file);
    }

    let readable_file = serialize_to_legacy_format(readable_files);

    Ok((results, readable_file))
}

#[cfg(test)]
mod tests {
    use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
    use base64::Engine;
    use flate2::Compression;
    use std::io::Write;

    use super::*;
    use insta::{assert_yaml_snapshot, glob};

    fn file_into_bytes(filename: &str) -> Vec<u8> {
        let upload = std::fs::read(filename).unwrap();
        let mut encoder = flate2::write::ZlibEncoder::new(Vec::new(), Compression::default());
        encoder.write_all(&upload).unwrap();
        let compressed = encoder.finish().unwrap();
        let base64_data = BASE64_STANDARD.encode(compressed);
        let upload_json = format!(
            r#"{{"network": [], "test_results_files": [{{"filename": "{}", "format": "base64+compressed", "data": "{}"}}]}}"#,
            filename.split('/').last().unwrap(),
            base64_data,
        );
        upload_json.into()
    }

    #[test]
    fn test_parse_raw_upload_success() {
        glob!("../tests", "*.xml", |path| {
            let upload_json = file_into_bytes(path.to_str().unwrap());
            let result = parse_raw_upload(&upload_json);
            match result {
                Ok((results, _)) => assert_yaml_snapshot!(results),
                Err(e) => {
                    assert_yaml_snapshot!(e.to_string());
                }
            }
        });
    }
}
