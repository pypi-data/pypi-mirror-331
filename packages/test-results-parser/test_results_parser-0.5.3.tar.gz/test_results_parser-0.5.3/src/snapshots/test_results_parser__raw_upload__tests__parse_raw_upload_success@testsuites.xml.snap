---
source: src/raw_upload.rs
expression: results
input_file: tests/testsuites.xml
---
- framework: ~
  testruns: []
